include/kernel.h

```diff
@@ -5861,6 +5861,29 @@ extern void k_sys_runtime_stats_enable(void);
  */
 extern void k_sys_runtime_stats_disable(void);
 
+/**
+ * @brief Wait for an expression to return true with a timeout
+ *
+ * Spin on an expression with a timeout and optional delay between iterations
+ *
+ * Commonly needed when waiting on hardware to complete an asynchronous
+ * request to read/write/initialize/reset, but useful for any expression.
+ *
+ * @param expr Trueth expression you wish to spin on, ex: XYZREG & XYZREG_EN
+ * @param timeout Timeout to wait for, ex: K_MSEC(100)
+ * @param delay_stmt Delay statement, ex: k_yield(), k_msleep(1) or k_busy_wait(1)
+ *
+ * @retval expr As a boolean return, if false then it has timed out.
+ */
+#define k_expr_wait(expr, timeout, delay_stmt)                                                     \
```

Comment:
why the k_ namespace and not just as a utility macro in utils.h?

Reply:
No particular reason, but it seemed like a nice spot for this (ie: k_sem/k_poll/etc) as its a really common pattern. Happy to move it anywhere that's most sensible.